nmea_parser.c

```diff
@@ -1,191 +1,315 @@
+	int i,j,k;
+
+	for (i=j=0;((ptr[i]!=0) && (ptr[i]!='*'));++i) j^=ptr[i];
+	k=j&0xf;
+	if (ptr[i]!='*') return 0;
+	if (k<10) {
+		if (ptr[i+2] != '0'+k) return 0;
+	} else { 
+		if (ptr[i+2] != 'A'+k-10) return 0;
+	}
+	j=j>>4;
+	if (j<10) {
+		if (ptr[i+1] != '0'+j) return 0;
+	} else {
+		if (ptr[i+1] != 'A'+j-10) return 0;
+	}
+	return i;
+}
+
+void checksum (char *ptr) {
+	int i,j,k;
+
+	for (i=1,j=0,k=strlen(ptr);i<k;++i) j^=ptr[i];
+	sprintf(ptr+k,"*%02X",j);
+}
+
+inline void tx_mesg (char *client_message, int xcsoar_sock) {
+
+	checksum(client_message);
+	strcat(client_message,"\r\n");
+	if (send(xcsoar_sock, client_message, strlen(client_message), 0) < 0) 
+		fprintf(stderr, "send failed\n");
+}
+
+void parse_NMEA_sensor(char* message, t_sensor_context* sensors, int xcsoar_sock)
 { 
-	char *val;
-	char buffer[2001];
-	char delimiter[]=",*";
-	char *ptr=NULL;
-	
+	static char buffer[2001];
+	char del1[2]="$",del2[2]=",";
+	char *ptr=NULL,*ptr2,*last;
+	char tick=1;
+	char mesg[120];
+	int i;
+	static unsigned char counter=255;
+
 	// copy string and initialize strtok function
-	strncpy(buffer, message, strlen(message));
-	ptr = strtok(buffer, delimiter);	
-	
-	while (ptr != NULL)
-	{
-		
-		switch (*ptr)
-		{
-			case '$':
-			// skip start of NMEA sentence
-			break;
-			
-			case 'E':
-			// TE vario value
-			// get next value
-			ptr = strtok(NULL, delimiter);
-			val = (char *) malloc(strlen(ptr));
-			strncpy(val,ptr,strlen(ptr));
-			sensors->e = atof(val);
-			break;
-			
-			case 'Q':
-			// Airspeed value
-			// get next value
-			ptr = strtok(NULL, delimiter);
-			val = (char *) malloc(strlen(ptr));
-			strncpy(val,ptr,strlen(ptr));
-			sensors->q = atof(val);
-			break;
-			default:
-			break;
-		}
-		// get next part of string
-		ptr = strtok(NULL, delimiter);
+	if (status) {
+		if (strchr(buffer,'$')==NULL) counter++;
+		if (counter==0) {
+			strcpy(mesg,"$POV,?");
+			if (status&0x01) strcat(mesg,",POL");
+			if (status&0x02) strcat(mesg,",RPO");
+			if (status&0x04) strcat(mesg,",MC");
+			if (status&0x08) strcat(mesg,",WL");
+			if (status&0x10) strcat(mesg,",BU");
+			if (status&0x20) strcat(mesg,",VAR");
+			tx_mesg(mesg,xcsoar_sock);
+			printf ("%s\n",mesg);
+			counter=1;
+		} 
+	}
+	buffer[2000]=0;
+	if ((ptr2=strchr(buffer,'$'))!=NULL) {
+		for (i=0;ptr2[i]!=0;++i) buffer[i]=ptr2[i];
+		buffer[i]=0;
+	}
+	if (strlen(buffer)+strlen(message)>1999)
+		strcpy(buffer,message);
+	else strcpy(buffer+strlen(buffer),message); // Append the new sentence to the last
+	ptr = strtok(buffer, del1); // Tokenize about '$'
+	last=buffer;
+	while (ptr != NULL) {
+		last=ptr;
+		tick=1;
+		if ((*ptr=='P') && (*(ptr+1)=='O') && (*(ptr+2)=='V') && (*(ptr+3)==',')) { // If you begin the sentence correctly  
+			if (verify_checksum(ptr)) { // And checksum passes
+				tick=0;
+				ptr2=ptr+4; // Start looking after the $POV,
+				ptr+=strlen(ptr)+1; // Move ptr to the next POV
+				ptr2=strtok(ptr2,del2); // Tokenize about ','
```

Comment:
Are you sure you are using strtok() correctly? It keeps an internal state and it modifies the buffer it is working on.
Moreover: read_float_from_sentence() uses strtok(), too!

Reply:
I am pretty sure I am using it correctly.  I spent quite a bit of time trying to make sure I did everything correctly.  It is now 4:40AM here, and I should be asleep.  First, please review the CURRENT code instead of old code.  If you feel there is an actual problem, please let me know.  If you don't like the way I did something, like I used a function to generate a checksum and a different function to verify the checksum... Please don't complicate my life with this.  I am trying to address real problems which you brought up, and I have tested and verified that I have addressed each of these problems.

Reply:
Go ahead. Do what you want.